README.md
# Factorio-Wills-Spaceship-Repair
Will runs a spaceship repair business


A spaceship repair business where you need to launch expensive spaceship parts to space for customer orders. Aim is to make some profit before you go bankrupt with debt. Investors (viewers) make investments giving you some cash to spend in a market to make life easier, but require a large return on their investment after many hours. Their investment accures interest after the payback period.
Delivering spaceship parts in to space requires an available dry dock to be in space and for an order to be decrypted (researched). If the order is fulfilled witihn the time limit a time bonus/penalty payment will be made in coins to the silo. These coins can be spent in the market or processed in to coin shipment capsules and sent off via rocket to the bank.
Rocket silos and coin handling machines are infrequently randomly placed across the map and can not be manually made. Rocket silos are stage 1 of the Rocket Silo Construction mod and can not be mined. If destroyed they return to stage 1 and must be rebuilt.
Supporting workforce (above first player) has to be researched and then paid an hourly wage when they are on shift.
All high tech personal equipment must be brought through the market which is at spawn area.
Instant cash will either appear in the passive provider chest near spawn or via delivery pod if the Item Delivery Pod mod is enabled.

Designed for play in multiplayer as a streamer with viewers acting as investors.


Commands
------

- wills_spaceship_repair-add_investment = Add a new Investment using the command arguments: "Player Name" InvestmentAmount
- wills_spaceship_repair-write_order_data = Writes out details of all orders to the that players script-output folder.
- wills_spaceship_repair-write_investment_data = Writes out details of investments to the that players script-output folder.


Settings (Advanced)
-------

- Item delivery pod size cash values = When the optional mod Item Delivery Pod is being used the size of the delivery pod for instant cash is defined by this setting. It is a JSON table of the coinCost, shipSize, radius. The coin cost is the minimum value for the ship size. The shipSizes are, tiny, small, medium, large, modular. The radius is how close to the targeted player the item delivery pod will randomly fall within. An extra entry for shipSize 'modular-part' is also required which controls how quickly the modular item delivery pod size grows. Default value showing data structure:
'[ {"coinCost":0, "shipSize":"tiny", "radius":150}, {"coinCost":500, "shipSize":"small", "radius":100}, {"coinCost":1000, "shipSize":"medium", "radius":75}, {"coinCost":2500, "shipSize":"large", "radius":50}, {"coinCost":3000, "shipSize":"modular", "radius":50}, {"coinCost":500, "shipSize":"modular-part"} ]'


Credits
-----
Many graphics and recipes copied/inspired from Space Extension Mod (SpaceX) under its MIT licence.
Colored assembling machine graphics from Bobs Assembling Machines with permission.


Usage & Testing Notes
------
The mod may not test correctly with Creative Mode and this mod isn't supported.
Don't use with an island map.
Must be used with a new game, won't add to an existing save.



Notes on mod
----------
The coin machine type will be randomly selected with a ratio of 3 wooden coin chest to 1 iron coin chest. A rocket silos and coin chest will be somewhere within a 2,500 tile sqaure region with the corners of the 4 starting regions at spawn (0,0). There is a single steel chest coin machine around spawn by the market and coin machines are not affected by any module. When hovering over a rocket silo its inventory will be listed to that player as a tooltip, however, this tooltip may display oddly at particularly high or low zooms.
Rocket silo's have an output (result) inventory of 25,000 stacks, enough to hold the most valuable ship part completed at the maximum delivery bonus.

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 18.0.4
Date: 2020-xx-xx
  Features:
    - updated for Factorio 0.18 as-is
  Changes:
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 17.0.3
Date: 2019-07-27
  Features:
  Changes:
  Bugfixes:
    - corrected error with part chance number generation
  Info:
---------------------------------------------------------------------------------------------------
Version: 17.0.2
Date: 2019-07-26
  Features:
    - Added on-hover text over the rocket silo to list the silos output (result) inventory
    - If "Primary Player Name" setting is blank and the "Item Delivery Pod" mod is active a random online player will be targeted.
  Changes:
    - Red text is lighter to improve readability on stream
    - Order Audit records the number of parts launched out of the required number.
    - Rocket Silo's behaviour is set via the mod and no longer requires specific Rocket Silo Construction mod settings.
  Bugfixes:
    - Fix market price discrepencies
    - Fix Investments Total recalculation from mod setting changes or version updates
    - Order Gui Bonus text corrected for the final minute of each bonus time period.
    - Order Audit had duplicate entries for multi part orders.
    - A cancelled and then restarted order decryption could get multiple drydocks in to the "Decrypting" state. The decryption technology would decrypt all drrdocks in the "Decrypting" state at once.
  Info:
---------------------------------------------------------------------------------------------------
Version: 17.0.1
Date: 2019-07-19
  Features:
    - Add Investment Delete command. It updates all totals when ran.
    - Add Investment GUI condensing option to reduce the clutter of many small investments in the Investment GUI screen.
  Changes:
    - Coins and coin chests are now selectable in filters.
  Bugfixes:
    - Fix issue with Investments being wiped on mod upgrade.
    - Fix issue with Investment Dividend Multiplyer setting being changed not recalculating Overdraft limit
    - When placing player entites in the map raise the built event for other mods.
  Info:
    - Requires Rocket Silo Construciton mod 0.17.9 as this has a fix for mod placed rocket silo construction sites.
---------------------------------------------------------------------------------------------------
Version: 17.0.0
Date: 2019-07-10
  Features:
    - Initial Release
  Changes:
  Bugfixes:
  Info:

settings.lua
data:extend(
    {
        {
            name = "wills_spaceship_repair-investment_dividend_multiplier",
            type = "double-setting",
            default_value = 10,
            minimum_value = 0,
            setting_type = "runtime-global",
            order = "1001"
        },
        {
            name = "wills_spaceship_repair-investment_cash_multiplier",
            type = "double-setting",
            default_value = 1,
            minimum_value = 0,
            setting_type = "runtime-global",
            order = "1002"
        },
        {
            name = "wills_spaceship_repair-investment_maturity_minutes",
            type = "int-setting",
            default_value = 300,
            minimum_value = 0,
            setting_type = "runtime-global",
            order = "1003"
        },
        {
            name = "wills_spaceship_repair-matured_investment_dividend_hourly_interest",
            type = "double-setting",
            default_value = 2,
            minimum_value = 0,
            setting_type = "runtime-global",
            order = "1004"
        },
        {
            name = "wills_spaceship_repair-workforce_minute_wage",
            type = "int-setting",
            default_value = 10,
            minimum_value = 0,
            setting_type = "runtime-global",
            order = "1005"
        },
        {
            name = "wills_spaceship_repair-starting_debt_ceiling",
            type = "int-setting",
            default_value = 1000000,
            minimum_value = 0,
            setting_type = "runtime-global",
            order = "1006"
        },
        {
            name = "wills_spaceship_repair-profit_target",
            type = "int-setting",
            default_value = 50000,
            minimum_value = 0,
            setting_type = "runtime-global",
            order = "1007"
        },
        {
            name = "wills_spaceship_repair-profit_label",
            type = "string-setting",
            default_value = "Profit",
            setting_type = "runtime-global",
            order = "1008"
        },
        {
            name = "wills_spaceship_repair-primary_player_name",
            type = "string-setting",
            default_value = "",
            allow_blank = true,
            setting_type = "runtime-global",
            order = "1009"
        },
        {
            name = "wills_spaceship_repair-item_delivery_pod_size_cash_values",
            type = "string-setting",
            default_value = '[ {"coinCost":0, "shipSize":"tiny", "radius":150}, {"coinCost":500, "shipSize":"small", "radius":100}, {"coinCost":1000, "shipSize":"medium", "radius":75}, {"coinCost":2500, "shipSize":"large", "radius":50}, {"coinCost":3000, "shipSize":"modular", "radius":50}, {"coinCost":500, "shipSize":"modular-part"} ]',
            allow_blank = true,
            setting_type = "runtime-global",
            order = "1010"
        },
        {
            name = "wills_spaceship_repair-investment_gui_condensing_minutes",
            type = "int-setting",
            default_value = 0,
            minimum_value = 0,
            setting_type = "runtime-global",
            order = "1011"
        }
    }
)
